lib/Optimizer/Transforms/DecompositionPatterns.cpp

```diff
@@ -0,0 +1,922 @@
+  for (auto control : opControls) {
+    if (auto veq = dyn_cast<quake::VeqType>(control.getType())) {
+      if (!veq.hasSpecifiedSize())
+        return failure();
+      numControls += veq.getSize();
+      continue;
+    }
+    numControls += 1;
+  }
+
+  return numControls == requiredNumControls ? success() : failure();
+}
+
+/// Check whether the operation has the correct number of controls. This
+/// function take as input a mutable array reference, `controls`, which must
+/// have the size equal to the number of controls. If the operation has `veq`s
+/// as controls, split those into single qubit references.
+///
+/// Note: This function assumes that the operation has already been tested for
+/// reference semantics.
+LogicalResult checkAndExtractControls(quake::OperatorInterface op,
+                                      MutableArrayRef<Value> controls,
+                                      PatternRewriter &rewriter) {
+  if (failed(checkNumControls(op, controls.size())))
+    return failure();
+
+  size_t controlIndex = 0;
+  for (Value control : op.getControls()) {
+    if (auto veq = dyn_cast<quake::VeqType>(control.getType())) {
+      for (size_t i = 0, end = veq.getSize(); i < end; ++i) {
+        Value index = createConstant(op.getLoc(), i, rewriter);
+        Value qref =
+            rewriter.create<quake::ExtractRefOp>(op.getLoc(), control, index);
+        controls[controlIndex] = qref;
+        controlIndex += 1;
+      }
+    } else {
+      controls[controlIndex] = control;
+      controlIndex += 1;
+    }
+  }
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// HOp decompositions
+//===----------------------------------------------------------------------===//
+
+// quake.h target
+// ───────────────────────────────────
+// quake.phased_rx(π/2, π/2) target
+// quake.phased_rx(π, 0) target
+struct HToPhasedRx : public OpRewritePattern<quake::HOp> {
+  using OpRewritePattern<quake::HOp>::OpRewritePattern;
+
+  void initialize() { setDebugName("HToPhasedRx"); }
+
+  LogicalResult matchAndRewrite(quake::HOp op,
+                                PatternRewriter &rewriter) const override {
+    if (!op.getControls().empty())
+      return failure();
+    if (!quake::isAllReferences(op))
+      return failure();
+
+    // Op info
+    Location loc = op->getLoc();
+    Value target = op.getTarget();
+
+    // Necessary/Helpful constants
+    ValueRange noControls;
+    Value zero = createConstant(loc, 0.0, rewriter.getF64Type(), rewriter);
+    Value pi = createConstant(loc, M_PI, rewriter.getF64Type(), rewriter);
+    Value pi_2 = createConstant(loc, M_PI_2, rewriter.getF64Type(), rewriter);
+
+    std::array<Value, 2> parameters = {pi_2, pi_2};
+    rewriter.create<quake::PhasedRxOp>(loc, parameters, noControls, target);
+    parameters[0] = pi;
+    parameters[1] = zero;
+    rewriter.create<quake::PhasedRxOp>(loc, parameters, noControls, target);
```

Comment:
MLIR question - can you replace the create + erase with `replaceOpWithNewOp(...)`?

Reply:
We could, but this is more useful with operations that return values. Here, wouldn't make much of a difference.